Blueototh: UUID: add string to uuid

chengkai15 opened the pull request:
# Bluetooth: UUID: add string to UUID conversion function

## Summary
This PR adds a new `bt_str_to_uuid()` function to convert string representations to Bluetooth UUID structures, complementing the existing `bt_uuid_to_str()` function. This provides bidirectional conversion between UUID structures and their string representations.

## Changes
- **New API**: Add `bt_str_to_uuid()` function in `uuid.h` header
- **Implementation**: Implement string parsing for all UUID formats in `uuid.c`
- **Format Support**: Handle 16-bit, 32-bit, and 128-bit UUID string formats

## Technical Details
- **16-bit UUID**: Parses 4-character hexadecimal strings (e.g., "1800")
- **32-bit UUID**: Parses 8-character hexadecimal strings (e.g., "00001800")  
- **128-bit UUID**: Parses standard UUID format (e.g., "00001800-0000-1000-8000-00805f9b34fb")
- **Validation**: Returns `true` on success, `false` on invalid input format
- **Error Handling**: Comprehensive format validation with proper return codes

## Features
- **Bidirectional Conversion**: Complements existing `bt_uuid_to_str()` function
- **Flexible Input**: Supports all standard UUID string formats
- **Type Detection**: Automatically detects and sets appropriate UUID type (16/32/128-bit)
- **Robust Parsing**: Uses `sscanf` with exact format matching for reliable parsing

## Usage Example
```c
struct bt_uuid uuid;
const char *uuid_str = "1800";

if (bt_str_to_uuid(uuid_str, &uuid)) {
    // Successfully converted string to UUID
    // uuid.type will be BT_UUID_TYPE_16
    // BT_UUID_16(&uuid)->val will be 0x1800
}
```

## Testing
- Verified conversion of 16-bit, 32-bit, and 128-bit UUID strings
- Tested edge cases and invalid input formats
- Confirmed proper type detection and value assignment
- Validated round-trip conversion with `bt_uuid_to_str()`

## Impact
- **New Feature**: Adds missing string-to-UUID conversion capability
- **Backward Compatible**: No changes to existing APIs
- **Minimal Overhead**: Small footprint with comprehensive functionality
- **Utility Enhancement**: Useful for configuration, debugging, and dynamic UUID handling

This function fills an important gap in the Bluetooth UUID utilities, enabling applications to easily convert user-input or configuration strings into proper UUID structures for Bluetooth operations.

chengkai15 committed:
Bluetooth: UUID: add string to UUID conversion function

Implement bt_uuid_from_str() function to convert string representations
to Bluetooth UUID structures. This provides bidirectional conversion
between UUID structures and their string representations, complementing
the existing bt_uuid_to_str() function.

Features:
- Supports 16-bit UUID format (4 hex characters, e.g., "180d")
- Supports 32-bit UUID format (8 hex characters, e.g., "abcdef12")
- Supports 128-bit UUID standard format (with hyphens)
- Comprehensive input validation and error handling
- Automatic UUID type detection based on string length

The function automatically detects the UUID type (16/32/128-bit) based
on the input string length and format, setting the appropriate type in
the output UUID structure.

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

Thalley commented:
Please expand https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bluetooth/uuid to test this new function :)

alwa-nordic commented on include/zephyr/bluetooth/uuid.h:
```diff
@@ -5249,6 +5249,18 @@ bool bt_uuid_create(struct bt_uuid *uuid, const uint8_t *data, uint8_t data_len)
+ *
+ *  @return true if the string was valid and the UUID was successfully created.
+ */
+bool bt_str_to_uuid(const char *str, struct bt_uuid *uuid);
```
I suggest using `from` and also reversing the parameters. That way, we both get `bt_uuid` as a namespace, and we get `bt_uuid` as a `self` (to borrow terminology from Python).

```suggestion
bool bt_uuid_from_str(struct bt_uuid *uuid, const char *str);
```

alwa-nordic commented on include/zephyr/bluetooth/uuid.h:
```diff
@@ -5249,6 +5249,18 @@ bool bt_uuid_create(struct bt_uuid *uuid, const uint8_t *data, uint8_t data_len)
+ *  UUID can be in any format, 16-bit, 32-bit or 128-bit.
+ *
+ *  @param str pointer to string to convert
+ *  @param uuid pointer where to put converted UUID
```
We can also create a type compatible with all the subtypes and make a parsing function that selects the variant tag corresponding to the string format, or to the smallest type that can represent the BT UUID. Maybe one function for each of those options.

chengkai15 committed:
Bluetooth: Test: add unit tests for bt_str_to_uuid function

Add comprehensive unit tests for the new bt_str_to_uuid function
to ensure proper conversion of string representations to Bluetooth
UUID structures. Tests cover:

- 16-bit UUID string conversion (e.g., "180d")
- 32-bit UUID string conversion (e.g., "abcdef12")
- 128-bit UUID string conversion with standard format
- Invalid input cases (malformed strings, empty strings, wrong lengths)

Includes proper validation of converted UUID values against
expected results using bt_uuid_cmp().

Signed-off-by: Kai Cheng <chengkai@xiaomi.com>

chengkai15 commented:
> Please expand https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bluetooth/uuid to test this new function :)

I add it with https://github.com/zephyrproject-rtos/zephyr/pull/98784/commits/4ed0d20c004a23ba6dd891911a0d2567f2cdc95f

However, I'm not sure how to set up the test environment.

alwa-nordic commented:
> I'm not sure how to set up the test environment.

Try this:

```bash
/zephyrproject$ uv venv --clear && uv pip install -r zephyr/scripts/requirements.txt && uv run -- bash -c 'source zephyr/zephyr-env.sh && west twister -ciT zephyr/tests/bluetooth/uuid/'
```

Thalley commented on include/zephyr/bluetooth/uuid.h:
```diff
@@ -5249,6 +5249,27 @@ bool bt_uuid_create(struct bt_uuid *uuid, const uint8_t *data, uint8_t data_len)
+ *  UUID format, which could be 16 bit, 32 bit or 128 bit format. And the
+ *  uuid pointer must point to a bt_uuid large enough to hold the converted
+ *  UUID. If the string uuid is 16 bit, the converted bt_uuid will be
+ *  bt_uuid_16 type. If the string uuid is 32 bt uuid, the converted bt_uuid
```
```suggestion
 *  bt_uuid_16 type. If the string uuid is 32 bit uuid, the converted bt_uuid
```

or did you mean to use `BT`/Bluetooth?

chengkai15 commented on include/zephyr/bluetooth/uuid.h:
```diff
@@ -5249,6 +5249,27 @@ bool bt_uuid_create(struct bt_uuid *uuid, const uint8_t *data, uint8_t data_len)
+ *  For example: 128 bit format "00001101-0000-1000-8000-00805F9B34FB"
+ *
+ *  @param uuid pointer where to put converted UUID
+ *  @param str pointer to string to convert
```
updated

chengkai15 commented on include/zephyr/bluetooth/uuid.h:
```diff
@@ -5249,6 +5249,27 @@ bool bt_uuid_create(struct bt_uuid *uuid, const uint8_t *data, uint8_t data_len)
+ *
+ *  @return 0 if the string was converted successfully, otherwise negative error
+ */
+int bt_uuid_from_str(struct bt_uuid *uuid, const char *str);
```
updated

jhedberg commented on subsys/bluetooth/host/uuid.c:
```diff
@@ -130,3 +132,49 @@ void bt_uuid_to_str(const struct bt_uuid *uuid, char *str, size_t len)
 	}
 }
+
+int bt_uuid_from_str(const char *str, struct bt_uuid *uuid)
```
Seems a bit inconsistent with the `bt_uuid_create()` API return type. Also, there doesn't really seem to exist any other category of failure than "I can't interpret this as a UUID".
